File: src/Main/java/com/nishant/algorithms/DataStructures/Trie/Trie.java
class Trie {
	class TrieNode {
		public TrieNode[] children;
		public int count;

		public TrieNode() {
			children = new TrieNode[26];
		}
	}

	private TrieNode root;

	public Trie() {
		root = new TrieNode();
	}

	public void insert(String s) {
		TrieNode curr = root;
		char[] chars = s.toCharArray();

		for(int i = 0; i < chars.length; i ++) {
			int index = chars[i] - 'a';

			if(curr.children[index] == null) {
				curr.children[index] = new TrieNode();
			}

			curr = curr.children[index];
		}

		curr.count ++;
	}

	public boolean contains(String s) {
		TrieNode curr = root;
		char[] chars = s.toCharArray();

		for(int i = 0; i < chars.length; i ++) {
			int index = chars[i] - 'a';

			if(curr.children[index] == null) {
				return false;
			}

			curr = curr.children[index];
		}

		return curr.count > 0;
	}

	public static void main(String[] args) {
		Trie trie = new Trie();
		trie.insert("app");
		trie.insert("apple");
		trie.insert("ap");

		System.out.println(trie.contains("app"));
		System.out.println(trie.contains("appl"));
	}


}

File: src/main/python/com/nishant/datastructures/trie/trie.py
class trie:
    class node:
        def __init__(self):
            self.count = 0
            self.children = [None] * 26 

    def __init__(self):
        self.root = self.node()
        
    def insert(self, string):
        curr = self.root

        for i in range(len(string)):
            index = ord(string[i]) - ord('a')

            if curr.children[index] == None:
                curr.children[index] = self.node()

            curr = curr.children[index]

        curr.count = curr.count + 1

    def contains(self, string):
        curr = self.root

        for i in range(len(string)):
            index = ord(string[i]) - ord('a')

            if curr.children[index] == None:
               return False 

            curr = curr.children[index]

        return curr.count > 0

if __name__ == "__main__":
    t = trie()

    t.insert("apple")
    t.insert("app")

    print(t.contains("app"))
    print(t.contains("ap"))

